dbms/src/Storages/StorageKafka.cpp

```diff
@@ -0,0 +1,358 @@
+#include <Common/Exception.h>
+#include <Common/setThreadName.h>
+#include <DataStreams/FormatFactory.h>
+#include <DataStreams/IProfilingBlockInputStream.h>
+#include <DataStreams/LimitBlockInputStream.h>
+#include <DataStreams/copyData.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/InterpreterInsertQuery.h>
+#include <Parsers/ASTExpressionList.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTInsertQuery.h>
+#include <Storages/StorageKafka.h>
+#include <common/logger_useful.h>
+
+#include <rdkafka.h>
+
+#include <thread>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_DATA;
+    extern const int UNKNOWN_EXCEPTION;
+    extern const int CANNOT_READ_FROM_ISTREAM;
+}
+
+/// How long to wait for a single message (applies to each individual message)
+static const auto READ_POLL_MS       = 1 * 1000;
+static const auto CLEANUP_TIMEOUT_MS = 2 * 1000;
+
+/// How many messages to pull out of internal queue at once
+static const auto BATCH_SIZE_MAX     = 16UL;
+
+class ReadBufferFromKafkaConsumer : public ReadBuffer
+{
+    using Messages = std::vector<rd_kafka_message_t *>;
+
+    rd_kafka_t * consumer;
+    Messages messages;
+    Messages::iterator current;
+    Messages::iterator end;
+    Poco::Logger * log;
+
+    bool nextImpl() override
+    {
+        if (current == end)
+        {
+            // Fetch next batch of messages
+            bool res = fetchMessages();
+            if (!res)
+            {
+                LOG_ERROR(log, "Consumer error: " << rd_kafka_err2str(rd_kafka_last_error()));
+                return false;
+            }
+
+            // No error, but no messages read
+            if (current == end)
+                return false;
+        }
+
+        // Process next buffered message
+        rd_kafka_message_t * msg = *(current++);
+        if (msg->err)
+        {
+            if (msg->err != RD_KAFKA_RESP_ERR__PARTITION_EOF)
+                LOG_ERROR(log, "Consumer error: " << rd_kafka_err2str(msg->err) << " " << rd_kafka_message_errstr(msg));
+            return false;
+        }
+
+        // Consume message and mark the topic/partition offset
+        BufferBase::set(reinterpret_cast<char *>(msg->payload), msg->len, 0);
+        auto err = rd_kafka_offset_store(msg->rkt, msg->partition, msg->offset);
```

Comment:
Ideally this should just accumulate topic / partition offset in a separate table, and store it in `readSuffixImpl` or event later (when the stream chain finishes and the read data is safely written to the destination).